hazelcast/src/main/java/com/hazelcast/cache/impl/DeferredValue.java

```diff
@@ -0,0 +1,286 @@
+
+import com.hazelcast.nio.serialization.Data;
+import com.hazelcast.spi.serialization.SerializationService;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Thread-safe holder of value and/or its serialized form.
+ *
+ * @param <V> the type of value
+ */
+public final class DeferredValue<V> {
+
+    private static final DeferredValue NULL_VALUE;
+
+    private volatile Data serializedValue;
+    private volatile V value;
+    private volatile boolean valueExists;
+    private volatile boolean serializedValueExists;
+
+    static {
+        DeferredValue nullValue = new DeferredValue();
+        nullValue.valueExists = true;
+        nullValue.serializedValueExists = true;
+        NULL_VALUE = nullValue;
+    }
+
+    private DeferredValue() {
+    }
+
+    // get-or-deserialize-and-get
+    public V get(SerializationService serializationService) {
+        if (!valueExists) {
+            // it's ok to deserialize twice in case of race
+            assert serializationService != null;
+            value = serializationService.toObject(serializedValue);
+            valueExists = true;
+        }
+        return value;
+    }
+
+    public Data getSerializedValue(SerializationService serializationService) {
+        if (!serializedValueExists) {
+            assert serializationService != null;
+            serializedValue = serializationService.toData(value);
+            serializedValueExists = true;
+        }
+        return serializedValue;
+    }
+
+    // returns a new DeferredValue representing the same value as this
+    public DeferredValue<V> shallowCopy() {
+        if (this == NULL_VALUE) {
+            return NULL_VALUE;
+        }
+        DeferredValue<V> copy = new DeferredValue<V>();
+        if (serializedValueExists) {
+            copy.serializedValueExists = true;
+            copy.serializedValue = serializedValue;
+        }
+        if (valueExists) {
+            copy.valueExists = true;
+            copy.value = value;
+        }
+        return copy;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        DeferredValue<?> deferredValue = (DeferredValue<?>) o;
+
+        if (valueExists && deferredValue.valueExists) {
+            return value != null ? value.equals(deferredValue.value) : deferredValue.value == null;
+        }
+        if (serializedValueExists && deferredValue.serializedValueExists) {
+            return serializedValue != null ? serializedValue.equals(deferredValue.serializedValue)
+                    : deferredValue.serializedValue == null;
+        }
+        throw new IllegalArgumentException("Cannot compare serialized vs deserialized value");
```

Comment:
Can this create an issue if we add a new listener configuration to the cache config which was deserialised (thus containing only serialised listener configurations)?

Reply:
`CacheConfig.getListenerConfigurations().add(...)` is safe to use, however `CacheConfig.addCacheEntryListenerConfiguration` may hit this issue, as it adds the value directly in the set of `DeferredValue`s.